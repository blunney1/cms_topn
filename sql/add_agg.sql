--
--Testing cms_topn_add_agg function of the extension 
--

--prepare tables for aggregates
create table numbers (
	int_column int
);

INSERT INTO numbers SELECT 0 FROM generate_series(1,2000);
INSERT INTO numbers SELECT 1 FROM generate_series(1,100);
INSERT INTO numbers SELECT 2 FROM generate_series(1,1500);
INSERT INTO numbers SELECT 3 FROM generate_series(1,30);
INSERT INTO numbers SELECT 4 FROM generate_series(1,400000);
INSERT INTO numbers SELECT 5 FROM generate_series(1,60000);
INSERT INTO numbers SELECT NULL FROM generate_series(1,50);

create table strings (
	text_column text
);

INSERT INTO strings SELECT '0' FROM generate_series(1,2);
INSERT INTO strings SELECT '1' FROM generate_series(1,1);
INSERT INTO strings SELECT '2' FROM generate_series(1,150000);
INSERT INTO strings SELECT '3' FROM generate_series(1,200);
INSERT INTO strings SELECT '4' FROM generate_series(1,60);
INSERT INTO strings SELECT '5' FROM generate_series(1,700000);
INSERT INTO strings SELECT NULL FROM generate_series(1,300);

--check errors for unproper parameters
SELECT cms_topn_add_agg(int_column, 0) FROM numbers;
SELECT cms_topn_add_agg(int_column, -1, 0.2, 0.9) FROM numbers;
SELECT cms_topn_add_agg(int_column, 1, 2, 0.9) FROM numbers;
SELECT cms_topn_add_agg(int_column, 3, 0.2, 1.5) FROM numbers;

--check aggregates for fixed size types like integer with default parameters
SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers WHERE int_column < 0;
SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers WHERE int_column < 1;
SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers WHERE int_column < 2;
SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers WHERE int_column < 3;
SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers WHERE int_column < 4;
SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers WHERE int_column < 5;
SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers WHERE int_column < 6;
SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers;
SELECT cms_topn_info(cms_topn_add_agg(int_column, 3)) FROM numbers;

--check aggregates for variable length types like text
SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings WHERE text_column < '0';
SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings WHERE text_column < '1';
SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings WHERE text_column < '2';
SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings WHERE text_column < '3';
SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings WHERE text_column < '4';
SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings WHERE text_column < '5';
SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings WHERE text_column < '6';
SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings;
SELECT cms_topn_info(cms_topn_add_agg(text_column, 4)) FROM strings;
