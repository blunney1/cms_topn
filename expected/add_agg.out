--
--Testing cms_topn_add_agg function of the extension 
--
--prepare tables for aggregates
create table numbers (
	int_column int
);
INSERT INTO numbers SELECT 0 FROM generate_series(1,200);
INSERT INTO numbers SELECT 1 FROM generate_series(1,10);
INSERT INTO numbers SELECT 2 FROM generate_series(1,150);
INSERT INTO numbers SELECT 3 FROM generate_series(1,3);
INSERT INTO numbers SELECT 4 FROM generate_series(1,40000);
INSERT INTO numbers SELECT 5 FROM generate_series(1,6000);
INSERT INTO numbers SELECT NULL FROM generate_series(1,5);
create table strings (
	text_column text
);
INSERT INTO strings SELECT '0' FROM generate_series(1,2);
INSERT INTO strings SELECT '1' FROM generate_series(1,1);
INSERT INTO strings SELECT '2' FROM generate_series(1,15000);
INSERT INTO strings SELECT '3' FROM generate_series(1,20);
INSERT INTO strings SELECT '4' FROM generate_series(1,6);
INSERT INTO strings SELECT '5' FROM generate_series(1,70000);
INSERT INTO strings SELECT NULL FROM generate_series(1,30);
--check errors for unproper parameters
SELECT cms_topn_add_agg(int_column, 0) FROM numbers;
ERROR:  invalid parameters for cms_topn
HINT:  Number of top items has to be positive
SELECT cms_topn_add_agg(int_column, -1, 0.2, 0.9) FROM numbers;
ERROR:  invalid parameters for cms_topn
HINT:  Number of top items has to be positive
SELECT cms_topn_add_agg(int_column, 1, 2, 0.9) FROM numbers;
ERROR:  invalid parameters for cms_topn
HINT:  Error bound has to be between 0 and 1
SELECT cms_topn_add_agg(int_column, 3, 0.2, 1.5) FROM numbers;
ERROR:  invalid parameters for cms_topn
HINT:  Confidence interval has to be between 0 and 1
--check aggregates for fixed size types like integer with default parameters
SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers WHERE int_column < 0;
 topn 
------
(0 rows)

SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers WHERE int_column < 1;
  topn   
---------
 (0,200)
(1 row)

SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers WHERE int_column < 2;
  topn   
---------
 (0,200)
 (1,10)
(2 rows)

SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers WHERE int_column < 3;
  topn   
---------
 (0,200)
 (2,150)
 (1,10)
(3 rows)

SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers WHERE int_column < 4;
  topn   
---------
 (0,200)
 (2,150)
 (1,10)
(3 rows)

SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers WHERE int_column < 5;
   topn    
-----------
 (4,40000)
 (0,200)
 (2,150)
(3 rows)

SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers WHERE int_column < 6;
   topn    
-----------
 (4,40000)
 (5,6000)
 (0,200)
(3 rows)

SELECT topn(cms_topn_add_agg(int_column, 3), NULL::integer) FROM numbers;
   topn    
-----------
 (4,40000)
 (5,6000)
 (0,200)
(3 rows)

SELECT cms_topn_info(cms_topn_add_agg(int_column, 3)) FROM numbers;
                    cms_topn_info                    
-----------------------------------------------------
 Sketch depth = 5, Sketch width = 2719, Size = 106kB
(1 row)

--check aggregates for variable length types like text
SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings WHERE text_column < '0';
 topn 
------
(0 rows)

SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings WHERE text_column < '1';
 topn  
-------
 (0,2)
(1 row)

SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings WHERE text_column < '2';
 topn  
-------
 (0,2)
 (1,1)
(2 rows)

SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings WHERE text_column < '3';
   topn    
-----------
 (2,15000)
 (0,2)
 (1,1)
(3 rows)

SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings WHERE text_column < '4';
   topn    
-----------
 (2,15000)
 (3,20)
 (0,2)
 (1,1)
(4 rows)

SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings WHERE text_column < '5';
   topn    
-----------
 (2,15000)
 (3,20)
 (4,6)
 (0,2)
(4 rows)

SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings WHERE text_column < '6';
   topn    
-----------
 (5,70000)
 (2,15000)
 (3,20)
 (4,6)
(4 rows)

SELECT topn(cms_topn_add_agg(text_column, 4, 0.01, 0.09), NULL::text) FROM strings;
   topn    
-----------
 (5,70000)
 (2,15000)
 (3,20)
 (4,6)
(4 rows)

SELECT cms_topn_info(cms_topn_add_agg(text_column, 4)) FROM strings;
                    cms_topn_info                    
-----------------------------------------------------
 Sketch depth = 5, Sketch width = 2719, Size = 106kB
(1 row)

